### PR TITLE
[Feature] Support Multi-task.

### PR DESCRIPTION
**Pull request by piercus**
### Multi-tasking 

This is inspired from https://github.com/open-mmlab/mmclassification/pull/675 and https://github.com/open-mmlab/mmclassification/pull/808

### Inhomogenous dataset

One of the main change compared to #675 is to take into account some inhomogenous dataset.

By inhomogenous dataset, i mean, some sample may have ground truth value for one task, but not for another task, and we need to use mask in the mult-task training logic in order to modulate which sample to take into account with which loss.

### 1.x

We tried to follow 1.X standard as much as possible

### Made with open-mmlab in mind, not only mmcls

The main idea is to provide a generic and flexible multi-task toolbox.

This is for mmclassification, but we tried to limit the deps between the MultiTask Implementation and mmcls, which is mostly related to the data_sample types that can be handled.

We hope this can be discussed further, and we think making a more generic version of it may be part of another PR.

## Modification

Modifications are : 
* Dataset `MultiTaskDataset` + unit tests
* Pipeline `FormatMultiTaskLabels` + unit tests
* Metric `MultiTaskMetric` + unit tests
* Head `MultiTaskHead` + unit tests
* Structure `MultiTaskDataSample` + unit tests

## BC-breaking (Optional)

No

### Simple use case

The basic use case here it to classify multiple classes using the same neural network.

This can be useful if you want to "superclass" ImageNet for example (see https://robustness.readthedocs.io/en/latest/example_usage/custom_imagenet.html)

### Nested use case

If you have 4 classification task, 2 about classifying the background, and 2 about classiying the foreground, you can decide to organize your task like 
```
dict(type= 'MultiTaskHead', task_heads={
  "foreground": dict(type= 'LinearMultiTaskHead', task_heads={
    "foreground0": dict(type= 'LinearClsHead', num_classes=3),
    "foreground1": dict(type= 'LinearClsHead', num_classes=3)
  }),
  "background": dict(type= 'LinearMultiTaskHead', task_heads={
    "background0": dict(type= 'LinearClsHead', num_classes=3),
    "background1": dict(type= 'LinearClsHead', num_classes=3)
  })
})
```

By doing this, you can share layers between some tasks

## Checklist

**Before PR**:

- [x] Pre-commit or other linting tools are used to fix the potential lint issues.
- [x] Bug fixes are fully covered by unit tests, the case that causes the bug should be added in the unit tests.
- [x] The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
- [x] The documentation has been modified accordingly, like docstring
- [ ] example tutorials not done yet.

**After PR**:

- [ ] If the modification has potential influence on downstream or other related projects, this PR should be tested with those projects, like MMDet or MMSeg.
- [ ] CLA has been signed and all committers have signed the CLA in this PR.

### CODE REVIEW

**Comment by Ezra-Yu** [comment]
Are there any multi-task open-source datasets to test?

**Comment by piercus** [comment]
@Ezra-Yu 

MEDIC seems to be a good choice : https://crisisnlp.qcri.org/medic/index.html

**Comment by piercus** [comment]
@Ezra-Yu We have added an example config in `configs/mobilenet_v2/mobilenet_multi_task.py`

**Comment by mzr1996** [comment]
Good.
I will spend time reviewing it these days.

**Comment by piercus** [comment]
@mzr1996 an eval_mask has been added please share your feedback

File: mmcls/models/heads/multi_task_head.py [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+                task_sample.set_field(
+                    task_name in data_sample,
+                    'eval_mask',
+                    field_type='metainfo')
+                if task_name in data_sample:
+                    data_sample.get(task_name).update(task_sample)
+                else:
+                    data_sample.set_field(task_sample, task_name)
+
+        return data_samples
```
Comment (mzr1996) [inline mmcls/models/heads/multi_task_head.py:142]: How about using the below method? Then we don't need to modify the `ClsHead` and I have tested to train and do validation it on a demo dataset.
```suggestion
        predictions_dict = dict()

        for task_name, head in self.task_heads.items():
            task_samples = head.predict(feats)
            batch_size = len(task_samples)
            predictions_dict[task_name] = task_samples

        if data_samples is None:
            data_samples = [MultiTaskDataSample() for _ in range(batch_size)]

        for task_name, task_samples in predictions_dict.items():
            for data_sample, task_sample in zip(data_samples, task_samples):
                task_sample.set_field(
                    task_name in data_sample.tasks,
                    'eval_mask',
                    field_type='metainfo')

                if task_name in data_sample.tasks:
                    data_sample.get(task_name).update(task_sample)
                else:
                    data_sample.set_field(task_sample, task_name)

        return data_samples
```

**Review by mzr1996 — CHANGES_REQUESTED** [review]

**Review by mzr1996 — APPROVED** [review]

**Comment by piercus** [comment]
Thanks @marouaneamz for your work on this ! This has been a great internship :-)
